mitiq/vd/vd_utils.py

```diff
@@ -51,3 +55,137 @@ def map_for_grid_qubits(qu: cirq.Qid) -> cirq.Qid:
             new_circuit += circuit.transform_qubits(map_for_grid_qubits)
 
     return new_circuit
+
+
+def _apply_cyclic_system_permutation(
+    matrix: NDArray[np.complex128], N_qubits: int, M: int = 2
+) -> NDArray[np.complex128]:
+    """
+    Function that shifts the rows of a matrix or vector in such a way,
+    that each of the M registers of N_qubit qubits are shifted cyclically.
+    The implementation is identical to left multiplication with repeated swap
+    gates, however this optimisation in considerably faster.
+
+    Args:
+        matrix: The matrix or vector that should be shifted.
+        N_qubits: The number of qubits in each register.
+        M: The number of registers.
+
+    Returns:
+        The matrix or vector with the rows shifted cyclically.
+    """
+    matrix = np.array(matrix)
+
+    # determine the row permutation for the cyclic shift operation
+    permutation = [
+        j + i
+        for j in range(2**N_qubits)
+        for i in range(0, 2 ** (M * N_qubits), 2**N_qubits)
+    ]
+
+    # Some fancy index magic to permute the rows in O(n) time
+    # and space (n=2**(M*N_qubits))
+    idx = np.empty_like(permutation)
+    idx[permutation] = np.arange(len(permutation))
+
+    # allow to work with lists or arrays of np.ndarrays
+    if matrix.ndim == 2:
+        matrix = matrix[idx]
+    elif matrix.ndim == 3:
+        matrix[:] = matrix[:, idx]
+    else:
+        raise TypeError("matrix should be 2D or 3D ndarray")
+    return matrix
+
+
+def _apply_symmetric_observable(
+    matrix: NDArray[np.complex128],
+    N_qubits: int,
+    observable: Optional[Union[Observable, NDArray[np.complex128]]] = None,
+    M: int = 2,
+) -> NDArray[np.complex128]:
+    """
+    Function that applies a symmetric observable to a matrix or vector.
+    The observable is assumed to be the Z observable by default.
```

Comment:
1. Remove the sentence about $Z$ being default since it says so in the argument description.
```suggestion
    Function that applies a symmetric observable to a matrix or vector.
```
2. It's unclear what exactly the requirements are for `observable`. Does it need to be symmetric? If so, the code should check that and error if it's not. Is the observable on `N_qubits` or `N_qubits * M`?